### PR TITLE
Cursed implementation of allowing missing imports on wasmtime

### PR DESCRIPTION
**Pull request by pepyakin**
This is a quick and dirty implementation of the wasmtime counterpart of https://github.com/paritytech/substrate/pull/4550. This is aimed to unblock polkadot progress while we are waiting #4686.

The idea is that we employ parity-wasm to extract all the imports and do the resolving ourselves, then we collect all the missing function imports and create modules and stubs for them.

### CODE REVIEW

File: client/executor/wasmtime/src/runtime.rs [hunk 1]
```diff
@@ -115,9 +121,98 @@ pub fn create_instance(
...
+		let func_ty = match import_entry.external() {
+			parity_wasm::elements::External::Function(func_ty_idx) => {
+				let parity_wasm::elements::Type::Function(ref func_ty) =
+					types.get(*func_ty_idx as usize).ok_or_else(|| {
+						WasmError::Other(format!("corrupted module, type out of bounds"))
+					})?;
+				func_ty
+			}
+			_ => {
+				// The executor doesn't provide any non function imports!
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs]: When we switch to import memory, will this not "break" here? Why not ignore non function imports?

File: client/executor/wasmtime/src/runtime.rs [hunk 2]
```diff
@@ -115,9 +121,98 @@ pub fn create_instance(
...
+				return Err(WasmError::Other(format!(
+					"{}:{} is not a function",
+					import_entry.module(),
+					import_entry.field()
+				)));
+			}
+		};
+		let signature = convert_parity_wasm_signature(func_ty);
+
+		if import_entry.module() == "env" {
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs]: Can we have function exports that don't have the `module` `env`?

File: client/executor/wasmtime/src/runtime.rs [hunk 3]
```diff
@@ -115,9 +121,98 @@ pub fn create_instance(
...
+				.iter()
+				.find(|hf| hf.name() == import_entry.field())
+			{
+				if signature == hf.signature() {
+					continue;
+				}
+			}
+		}
+
+		// This import is not a function, not from env module, or doesn't have a corresponding host
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs]:
```suggestion
		// This import is a function, not from env module, or doesn't have a corresponding host
```

File: client/executor/wasmtime/src/runtime.rs [hunk 4]
```diff
@@ -115,9 +121,98 @@ pub fn create_instance(
...
+		// function, or the signature is not matching.
+		let sig = cranelift_ir_signature(signature, &call_conv);
+
+		missing_functions.insert(
+			import_entry.module().to_string(),
+			import_entry.field().to_string(),
+			sig
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs]:
```suggestion
			sig,
```

File: client/executor/wasmtime/src/runtime.rs [hunk 5]
```diff
@@ -130,8 +225,27 @@ fn create_compiled_unit(
...
+	} else {
+		// If there are in fact missing functions they will be detected at the instantiation time
+		// and the module will be rejected.
+		MissingFunctionStubs::new()
+	};
+
+	let env_missing_functions = missing_functions_stubs.stubs.remove("env").unwrap_or_else(|| Vec::new());
+	context.name_instance(
+		"env".to_owned(),
+		instantiate_env_module(global_exports, compiler, host_functions, env_missing_functions)?
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs]:
```suggestion
		instantiate_env_module(global_exports, compiler, host_functions, env_missing_functions)?,
```

File: client/executor/wasmtime/src/runtime.rs [hunk 6]
```diff
@@ -130,8 +225,27 @@ fn create_compiled_unit(
...
+
+	let mut missing_functions_stubs = if allow_missing_func_imports {
+		scan_missing_functions(code, host_functions)?
+	} else {
+		// If there are in fact missing functions they will be detected at the instantiation time
+		// and the module will be rejected.
+		MissingFunctionStubs::new()
+	};
+
+	let env_missing_functions = missing_functions_stubs.stubs.remove("env").unwrap_or_else(|| Vec::new());
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs]: Can we not do this in the loop below and just check:
```
if module == env {
    host_functions
} else {
    &[]
}
```

File: client/executor/wasmtime/src/runtime.rs [hunk 7]
```diff
@@ -231,6 +346,26 @@ fn instantiate_env_module(
 		finished_functions.push(trampoline);
 	}
 
+	for MissingFunction { name, sig } in missing_functions_stubs {
+		let sig = translate_signature(
+			sig,
+			pointer_type
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs]:
```suggestion
			pointer_type,
```

File: client/executor/wasmtime/src/runtime.rs (hunk 3 above)
Comment (pepyakin) [reply client/executor/wasmtime/src/runtime.rs]: Ah yeah you are right, I changed the condition but didn't change the comment.

File: client/executor/wasmtime/src/runtime.rs (hunk 2 above)
Comment (pepyakin) [reply client/executor/wasmtime/src/runtime.rs]: Yeah, we can! If I remember correctly, in Rust you'd use `#[link(wasm_import_module = "non_env")]` for that.

File: client/executor/wasmtime/src/runtime.rs (hunk 6 above)
Comment (pepyakin) [reply client/executor/wasmtime/src/runtime.rs]: The main point of doing that is not in host functions but rather in creating a new instance and giving it a name for `non-env` modules.

File: client/executor/wasmtime/src/runtime.rs (hunk 1 above)
Comment (pepyakin) [reply client/executor/wasmtime/src/runtime.rs]: Yeah, it will bail here! We can remove this in rely it to fail at the instantiation time.

**Pull request merged**